bin/varnishncsa/varnishncsa.c

```diff
@@ -1071,17 +1071,25 @@ dispatch_f(struct VSL_data *vsl, struct VSL_transaction * const pt[],
 			case SLT_RespHeader:
 				process_hdr(&CTX.watch_resphdr, b, e);
 				break;
+			case SLT_HitMiss:
+				CTX.hitmiss = "miss";
+				CTX.handling = "hitmiss";
+				break;
+			case SLT_HitPass:
+				CTX.hitmiss = "miss";
+				CTX.handling = "hitpass";
+				break;
 			case SLT_VCL_call:
 				if (!strcasecmp(b, "recv")) {
 					CTX.hitmiss = "-";
 					CTX.handling = "-";
 				} else if (!strcasecmp(b, "hit")) {
 					CTX.hitmiss = "hit";
 					CTX.handling = "hit";
-				} else if (!strcasecmp(b, "miss")) {
+				} else if (!strcasecmp(b, "miss") && strcmp(CTX.handling, "hitmiss")) {
 					CTX.hitmiss = "miss";
 					CTX.handling = "miss";
-				} else if (!strcasecmp(b, "pass")) {
+				} else if (!strcasecmp(b, "pass") && strcmp(CTX.handling, "hitpass")) {
 					CTX.hitmiss = "miss";
 					CTX.handling = "pass";
```

Comment:
Should we conditionally update only the hitmiss in these blocks?

In both `vcl_hit` and `vcl_miss` we can degrade to a pass, so technically hitmiss should be preserved (unless "-") while the handling should be unconditionally updated.

Reply:
I do not understand what you are trying to say. The intention is to preserve `hitmiss` and `hitpass` when we see a `call MISS` or `call PASS`. When `hitmiss` turns into a `pass`, the value ends up as `pass`. Is this what you are trying to say?

Reply:
So you are proposing exactly the same plus `purge` plus `reset`, just as _yet another_ different tag, which represents a superset of %handling which represents a superset of %hitmiss?

To clarify: 🟢 exists already 🟡 proposed by this PR 🔵 proposed by Dridi in addition

- 🟢 `hit`
- 🟢 `miss`
- 🟡 `hitmiss`
- 🟢 `pass` (no effective lookup)
- 🟡 `hitpass`
- 🟢 `pipe` (no effective lookup)
- 🔵 `purge`
- 🟢 `synth` (no effective lookup)
- 🔵 `reset` (client disconnected before a lookup-or-not decision was made)

I have not checked if `reset` can be implemented properly, at this point I would guess it might be equivalent to the existing `-`. So all in all, the proposal is to also add `purge`, fine.

And should we have a new %format? @gquintard says no, @dridi says yes, I don't care. But I do not think that tag inflation helps much here, I would just keep %handling.

Reply:
It seems like I forget to reply to some comments:

> What I had in mind was "%hitmiss/%handling" leading to for example "hitmiss/pass" after a return(pass) from vcl_miss in a hit-for-miss scenario.

Yes. How does this not work with the current patch? I added this to the test case: 143390be2

The `hitmiss` rightly gets turned into a `pass` because of a deliberate vcl return.

> What I mean by that is that a hit-for-pass transaction is handled the same as a regular pass transaction, whether you return(pass) from vcl_recv, vcl_hit, or vcl_miss. You essentially get a private fetch, this is how the request is handled. This is how I think %handling should be reported (and already is if I'm reading correctly).

😖 This patch proposes to report a difference between a regular `pass` and a `hitpass`. Before this patch, it was reported as `pass`, now as `hitpass`. That is intentional and I think it is helpful (unless we actually go for _yet another_ %format).

Reply:
wfm